File: modules/scale.c
#include <stdlib.h>
#include "utone.h"

int ut_scale_create(ut_scale **p)
{
    *p = malloc(sizeof(ut_scale));
    return UT_OK;
}

int ut_scale_destroy(ut_scale **p)
{
    free(*p);
    return UT_OK;
}

int ut_scale_init(ut_data *ut, ut_scale *p)
{
    p->min = -1;
    p->max = 1;
    return UT_OK;
}

int ut_scale_compute(ut_data *ut, ut_scale *p, UTFLOAT *in, UTFLOAT *out)
{
    *out =  *in * (p->max - p->min) + p->min;
    return UT_OK;
}

File: modules/data/scale.lua
sptbl["scale"] = {

    files = {
        module = "scale.c",
        header = "scale.h",
        example = "ex_scale.c",
    },

    func = {
        create = "ut_scale_create",
        destroy = "ut_scale_destroy",
        init = "ut_scale_init",
        compute = "ut_scale_compute",
    },

    params = {
        optional = {
            {
                name = "min",
                type = "UTFLOAT",
                description = "Minimum value to scale to.",
                default = 0
            },
            {
                name = "max",
                type = "UTFLOAT",
                description ="Maximum value to scale to.",
                default = 1
            },
        }
    },

    modtype = "module",

    description = [[Bipolar Scale

    This module scales from unipolar [0, 1] to another range defined by min and max.
]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "in",
            description = "Signal to be scaled."
        },
    },

    outputs = {
        {
            name = "out",
            description = "Scaled signal out"
        },
    }

}
